packages/syft/src/syft/core/tensor/smpc/share_tensor.py

```diff
@@ -118,23 +121,165 @@ def generate_przs(
+        Args:
+            share (Union[int, float, ShareTensor, np.ndarray, torch.Tensor]): value to check
+
+        Raises:
+            ValueError: if type is not supported
+        """
+        if isinstance(share, float):
+            raise ValueError("Type float not supported yet!")
+
+        if isinstance(share, np.ndarray) and not np.issubdtype(share.dtype, int):
+            raise ValueError(f"NPArray should have type int, but found {share.dtype}")
 
-    def __mul__(self, other: Any) -> "ShareTensor":
-        if isinstance(other, ShareTensor):
-            raise ValueError("Private Multiplication not yet implemented")
+        if isinstance(share, torch.Tensor) and torch.is_floating_point(share):
+            raise ValueError(f"Torch tensor should have type int, but found float")
+
+    def apply_function(
+        self, y: Union[int, float, torch.Tensor, np.ndarray, "ShareTensor"], op_str: str
+    ) -> "ShareTensor":
+        """Apply a given operation.
+
+        Args:
+            y (Union[int, float, torch.Tensor, np.ndarray, "ShareTensor"]): tensor to apply the operator.
+            op_str (str): Operator.
+
+        Returns:
+            ShareTensor: Result of the operation.
+        """
+
+        op = getattr(operator, op_str)
+        if isinstance(y, ShareTensor):
+            value = op(self.child, y.child)
         else:
-            return ShareTensor(value=self.child * other, rank=self.rank)
+            value = op(self.child, y)
+
+        res = self.copy_tensor()
+        res.child = value
+        return res
+
+    def add(
+        self, y: Union[int, float, torch.Tensor, np.ndarray, "ShareTensor"]
+    ) -> "ShareTensor":
+        """Apply the "add" operation between "self" and "y".
+
+        Args:
+            y (Union[int, float, torch.Tensor, np.ndarray, "ShareTensor"]): self + y
+
+        Returns:
+            ShareTensor. Result of the operation.
+        """
+        ShareTensor.sanity_check(y)
+        new_share = self.apply_function(y, "add")
+        return new_share
+
+    def sub(
+        self, y: Union[int, float, torch.Tensor, np.ndarray, "ShareTensor"]
+    ) -> "ShareTensor":
+        """Apply the "sub" operation between "self" and "y".
+
+        Args:
+            y (Union[int, float, torch.Tensor, np.ndarray, "ShareTensor"]): self - y
+
+        Returns:
+            ShareTensor. Result of the operation.
+        """
+
+        ShareTensor.sanity_check(y)
+        new_share = self.apply_function(y, "sub")
+        return new_share
+
+    def rsub(
+        self, y: Union[int, float, torch.Tensor, np.ndarray, "ShareTensor"]
+    ) -> "ShareTensor":
+        """Apply the "rsub" operation between "self" and "y"
+
+        Args:
+            y (Union[int, float, torch.Tensor, np.ndarray, "ShareTensor"]): y - self
+
+        Returns:
+            ShareTensor. Result of the operation.
+        """
+
+        ShareTensor.sanity_check(y)
+        new_self = self.mul(-1)
```

Comment:
Q: Any specific reason we did this instead of doing `y - sub`?

Reply:
I am not sure why that will work.
We got here because we called something like `y - x` (where x is a ShareTensor and possibly y is an integer).
Calling `y-sub` (sub == x?) means we are doing that recursively.